community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java

```diff
@@ -106,27 +116,101 @@ public NodeRelationshipCache( NumberArrayFactory arrayFactory, int denseNodeThre
      * @param nodeId node to increment relationship count for.
      * @return count after the increment.
      */
-    public int incrementCount( long nodeId )
+    public long incrementCount( long nodeId )
     {
         return incrementCount( array, nodeId, SPARSE_COUNT_OFFSET );
     }
 
-    void setCount( long nodeId, int count )
+    /**
+     * Should only be used by tests
+     */
+    void setCount( long nodeId, long count, Direction direction )
+    {
+        if ( isDense( nodeId ) )
+        {
+            long relGroupId = all48Bits( array, nodeId, SPARSE_ID_OFFSET );
+            relGroupCache.getAndSetCount( relGroupId, direction, count );
```

Comment:
Because the count we store is per relationship chain head. A dense node has 3 chain heads per type, which is why we need to store them in the relGroupCache for dense nodes.